ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java

```diff
@@ -0,0 +1,88 @@
+package io.ripc.protocol.tcp;
+
+import io.ripc.internal.Publishers;
+import org.reactivestreams.Publisher;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public abstract class TcpServer<R, W> {
+
+    @SuppressWarnings("rawtypes")
+    private LazyTcpHandler lazyHandler;
+    protected final TcpHandler<R, W> thisHandler;
+    protected final AtomicBoolean started;
+
+    protected TcpServer() {
+        lazyHandler = new LazyTcpHandler<>();
+        thisHandler = null;
+        this.started = new AtomicBoolean();
+    }
+
+    protected TcpServer(TcpHandler<R, W> handler) {
+        thisHandler = handler;
+        this.started = new AtomicBoolean();
+    }
+
+    public <RR, WW> TcpServer<RR, WW> intercept(TcpInterceptor<R, W, RR, WW> interceptor) {
+        if (null == thisHandler) {
+            @SuppressWarnings("unchecked")
+            TcpServer<RR, WW> toReturn = newServer(interceptor.intercept(lazyHandler));
+            toReturn.lazyHandler = lazyHandler;
+            return toReturn;
+        } else {
+            TcpServer<RR, WW> toReturn = newServer(interceptor.intercept(thisHandler));
+            toReturn.lazyHandler = lazyHandler;
+            return toReturn;
+        }
+    }
+
+    public final TcpServer<R, W> start(TcpHandler<R, W> handler) {
+        if (!started.compareAndSet(false, true)) {
+            throw new IllegalStateException("Server already started");
+        }
+
+        if (null == thisHandler) {
+            doStart(handler);
+        } else {
+            lazyHandler.start(handler);
+            doStart(thisHandler);
+        }
+        return this;
+    }
+
+    public final void startAndAwait(TcpHandler<R, W> handler) {
```

Comment:
Me & @benjchristensen talked about it and unless we model the server as a `Publisher` of `ServerState` which is `Start` and `Stop` it seems like an overkill since that would mean someone has to subscribe to the state observable to start the server. Since, there will only be two state `Start` and `Stop`, it sounds unnecessary to me to model it that way.
Are there any other usecases apart from await that you would see us gaining from if we move to returning a `Publisher` from `start()`?